2015: UTC offset

aosyatnik opened the pull request:
### Purpose

Link to YouTrack:
[MAGN-8993](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8993) Correct UTC offset for Solar Radiation workflow from Revit.

If I understood the issue correct... This is an old bug, that was first mentioned here in 2014 year: http://dynamobim.com/forums/topic/sunsettings-current/

SunSettings.CurrentDateTime as well as SunSettings.StartDateTime and SunSettings.EndDateTime return wrong datetime value. This value has some strange offset.

![image](https://cloud.githubusercontent.com/assets/8158404/12852534/151fb5e8-cc38-11e5-9bed-7660a33e3ed9.png)

For me this offset is 3 hours. But it depends on user's time zone. That's why MOHAMMAD RAHMANI (user from link above) had the same wrong offset, but for 5 hours.

This is bug of Revit API. It returns custom time - user time zone offset.
I found old API(2014) here: http://revitapisearch.com/html/e50880d2-cab4-3cb5-56c8-6a7faaa0725b.htm

It's said, that: "The date and time. The output value will be in Coordinated Universal Time (UTC), but input may be in local time as well." But I couldn't find any way to return local time.
So, I made some workaround:
I get users local timezone offset and add it to datetime, that provide Revit API.

Results:
![image](https://cloud.githubusercontent.com/assets/8158404/12852911/0b7478ba-cc3a-11e5-8cb0-9069daa674eb.png)
![image](https://cloud.githubusercontent.com/assets/8158404/12852998/5dda02e6-cc3a-11e5-81e6-f58d8607da25.png)

NB! Sometimes Sunsettings are null or wrong... For now problem is not found. And I can't reproduce it consistently.
![image](https://cloud.githubusercontent.com/assets/8158404/12853063/ad53778a-cc3a-11e5-8f6f-f6f79a3b6bea.png)

### Declarations

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.


### Reviewers

Could you try this fix in your time zone?
@mjkkirschner 
@ramramps 

### FYIs

@ikeough , I believe Ian knows more about it, then anyone else.

mjkkirschner commented on src/Libraries/RevitNodes/Elements/SunSettings.cs:
```diff
@@ -89,7 +89,7 @@ public double Azimuth
         /// </summary>
         public DateTime StartDateTime
```
can you modify the summaries to specify exactly what timezone the user can expect this time in, I.E -  "the start time of the solar study given in the local time of the solar study location" for all of these properties?

mjkkirschner commented on src/Libraries/RevitNodes/Elements/SunSettings.cs:
```diff
@@ -105,8 +105,27 @@ public DateTime EndDateTime
+        internal static DateTime TranslateTime(DateTime utc)
+        {
+            // Get user local hours offset.
+            var offset = TimeZoneInfo.Local.GetUtcOffset(DateTime.UtcNow).Hours;
```
this makes sense to me, what happens with daylight savings time? I don't think UTC has it, does the TimeZoneInfo take care of it?

mjkkirschner commented:
@aosyatnik a question, and one small change to the XML docs. Thanks for this fix :)

mjkkirschner commented:
also, I tested this on my machine on 2016 and it works for me.

aosyatnik commented:
Hi, @mjkkirschner , nice catch about day light saving...
I moved my clock to March 1. (no day light saving)

Then moved to March 27. (day light saving) - I had 1 additional hour.
So, I added check for Daylight Saving Time. Great, that `TimeZoneInfo` can check it with `IsDaylightSavingTime`!

Results:
![image](https://cloud.githubusercontent.com/assets/8158404/12947287/f8068c10-d002-11e5-8792-0baaa8488c70.png)
![image](https://cloud.githubusercontent.com/assets/8158404/12947359/618c30b8-d003-11e5-88b6-40e46441e409.png)

pbidenko commented:
@aosyatnik, this looks pretty good.
